### PR TITLE
Add support for binary file diffing

### PR DESCRIPTION
**Pull request by geomaster**
Added a new differ, `binary_diff`, which prints a single message if the files differ in size or contents. Added a parameter to `new_goldenfile` which allows the user to specify the differ to use. Updated tests and example usage.

This is a breaking change. All calls to `new_goldenfile` must include a new parameter, which should be `DifferType::Text` for regaining old behavior.

I thought it'd be best to make this a breaking change, seeing as there aren't a lot of users of this crate, and being explicit about the file type in code is better than e.g. detecting the file type based on the extension, and all other alternative ways to integrate this API felt weird. (Make a new method such as `new_goldenfile_with_differ` ? Then text diffing looks like the "default", and that default wasn't chosen, it just turned out to be the one that was first implemented, etc.) If you disagree, I'll be happy to make changes to this PR.

I'm currently using this library to verify in tests that there are no regressions regarding font rendering in my app. Since these tests produce image files (`.png`s), text diffing fails because these files aren't valid UTF-8 (and text diffing wouldn't have made sense anyhow). Bottomline is---I'm really liking `rust-goldenfile`. Thanks for building it!

### CODE REVIEW

**Comment by geomaster** [comment]
Whoops, looks like git decided not to track my new goldenfiles. Should be OK now.

**Comment by calder** [comment]
This looks great, sorry I didn't see it earlier! I'll do a full review this weekend.

File: src/differs.rs [hunk 1]
```diff
@@ -14,11 +14,45 @@ pub fn text_diff(old: &Path, new: &Path) {
...
+        .zip(new_reader.bytes())
+        .all(|(old_result, new_result)| {
+            let old_byte = old_result.expect(&format!("Error reading file: {:?}", old));
+            let new_byte = new_result.expect(&format!("Error reading file: {:?}", new));
+
+            old_byte == new_byte
+        });
+
+    if !no_difference {
+        panic!("Files differ in contents");
```
Comment (calder) [inline src/differs.rs]: Maybe add the index of the first byte they differ at too? You're call.

File: src/differs.rs [hunk 2]
```diff
@@ -14,11 +14,45 @@ pub fn text_diff(old: &Path, new: &Path) {
...
 
+/// Compare binary files. Print a "Files differ" message and panic on failure.
+pub fn binary_diff(old: &Path, new: &Path) {
+    let old_metadata = metadata(old)
+        .expect(&format!("Error querying for file metadata: {:?}", old));
+    let new_metadata = metadata(new)
+        .expect(&format!("Error querying for file metadata: {:?}", new));
+
+    if old_metadata.len() != new_metadata.len() {
+        panic!("Files differ in size");
```
Comment (calder) [inline src/differs.rs]: Add old and new file sizes to this message.

File: src/mint.rs [hunk 3]
```diff
@@ -25,6 +25,19 @@ pub struct Mint {
...
+/// and prints helpful output about what is different.
+pub enum DifferType {
+    /// A text differ compares unicode text files and prints colored diffs if
+    /// they do not match.
+    Text,
+    /// A binary differ compares any kinds of binary files, but does not print
+    /// any further info about the actual difference.
+    Binary
+}
+
```
Comment (calder) [inline src/mint.rs]: Why not just pass the diff function directly in the API (instead of having this enum at all)?

File: src/mint.rs [hunk 4]
```diff
@@ -42,13 +55,16 @@ impl Mint {
         return mint;
     }
 
-    /// Create a new goldenfile.
+    /// Create a new goldenfile, with a differ of the given type.
```
Comment (calder) [inline src/mint.rs]: Remove comma.

File: tests/basics.rs [hunk 5]
```diff
@@ -5,53 +5,100 @@ use std::fs::File;
 use std::io::Write;
 use std::path::Path;
 
-use goldenfile::Mint;
+use goldenfile::{Mint, DifferType};
 
-fn setup_file(path: &str, contents: &str) {
-    let path = Path::new(path);
+fn open_file(path: &str) -> File{
+let path = Path::new(path);
```
Comment (calder) [inline tests/basics.rs]: Weird formatting. Run `scripts/install-git-hooks` to automatically run a `cargo fmt` check during commit.

File: src/mint.rs [hunk 6]
```diff
@@ -42,13 +55,16 @@ impl Mint {
...
     ///
     /// The returned file is actually a temporary file, not the goldenfile
     /// itself. When the Mint goes out of scope, it will either check the temp
     /// file against the real goldenfile, or replace the real goldenfile based
     /// on the value of the `REGENERATE_GOLDENFILES` environment variable.
-    pub fn new_goldenfile<P: AsRef<Path>>(&mut self, path: P) -> Result<File> {
+    pub fn new_goldenfile<P: AsRef<Path>>(&mut self,
+                                          path: P,
+                                          differ_type: DifferType) -> Result<File> {
```
Comment (calder) [inline src/mint.rs]: Just take the differ function directly here and get rid of DifferType.

File: tests/basics.rs [hunk 7]
```diff
@@ -5,53 +5,100 @@ use std::fs::File;
...
 }
 
 #[test]
 fn basic_usage() {
-    setup_file("tests/goldenfiles/basic_usage1.txt", "Hello world!");
-    setup_file("tests/goldenfiles/basic_usage2.txt", "foobar");
+    setup_text_file("tests/goldenfiles/basic_usage1.txt", "Hello world!");
+    setup_text_file("tests/goldenfiles/basic_usage2.txt", "foobar");
+    setup_binary_file("tests/goldenfiles/basic_usage3.bin", b"BinaryData\x0f\x0c");
+    setup_binary_file("tests/goldenfiles/basic_usage4.bin", b"BinaryData2\xa0\xa0");
```
Comment (calder) [inline tests/basics.rs]: Rename basic_usage to zero_diff_text and move binary file testing out into a new zero_diff_binary function.

File: tests/basics.rs [hunk 8]
```diff
@@ -5,53 +5,100 @@ use std::fs::File;
...
 
 #[test]
 #[should_panic]
 fn positive_diff() {
-    setup_file("tests/goldenfiles/positive_diff1.txt", "Hello world!");
-    setup_file("tests/goldenfiles/positive_diff2.txt", "foobar");
+    setup_text_file("tests/goldenfiles/positive_diff1.txt", "Hello world!");
+    setup_text_file("tests/goldenfiles/positive_diff2.txt", "foobar");
+    setup_binary_file("tests/goldenfiles/basic_usage3.bin", b"BinaryData\x0f\x0c");
+    setup_binary_file("tests/goldenfiles/basic_usage4.bin", b"BinaryData2\xa0\xa0");
```
Comment (calder) [inline tests/basics.rs]: Remove binary files from this test (they're tested below).

File: tests/basics.rs [hunk 9]
```diff
@@ -5,53 +5,100 @@ use std::fs::File;
...
 #[test]
+#[should_panic]
+fn positive_diff_binary() {
+    setup_text_file("tests/goldenfiles/positive_diff1.txt", "Hello world!");
+    setup_text_file("tests/goldenfiles/positive_diff2.txt", "foobar");
+    setup_binary_file("tests/goldenfiles/basic_usage3.bin", b"BinaryData\x0f\x0c");
+    setup_binary_file("tests/goldenfiles/basic_usage4.bin", b"BinaryData2\xa0\xa0");
+
+    let mut mint = Mint::new("tests/goldenfiles");
+    let mut file1 = mint.new_goldenfile("positive_diff_binary1.txt", DifferType::Text).unwrap();
```
Comment (calder) [inline tests/basics.rs]: Remove text files from this test.

File: tests/basics.rs [hunk 10]
```diff
@@ -5,53 +5,100 @@ use std::fs::File;
...
 
     write!(file1, "Hello world!").unwrap();
     write!(file2, "foobar").unwrap();
+    file3.write(b"BinaryData\x0f\x0c").unwrap();
+    file4.write(b"BinaryData2\xa0\xa0").unwrap();
 }
 
 #[test]
 #[should_panic]
 fn positive_diff() {
```
Comment (calder) [inline tests/basics.rs]: Rename to positive_diff_text.

**Comment by calder** [comment]
@geomaster: I'd love to get this merged if you're up for it! Otherwise I'll close this PR in a few days if I don't get a response.

**Comment by calder** [comment]
@geomaster: Please re-open if you find some time to come back to this!

**Pull request closed without merging**